### PR TITLE
fix ipv4/ipv6 NXDOMAIN/NODATA confusion for synthesized local addresses

### PR DESCRIPTION
**Pull request by poettering**
Split out of #17535

### CODE REVIEW

File: src/shared/local-addresses.c [hunk 1]
```diff
@@ -93,6 +91,11 @@ int local_addresses(sd_netlink *context, int ifindex, int af, struct local_addre
                 if (flags & IFA_F_DEPRECATED)
                         continue;
 
+                if (!ret) {
+                        n_list++;
+                        continue;
+                }
+
```
Comment (yuwata) [inline src/shared/local-addresses.c]: We have several checks below. I think it is better to not skip them.

BTW, this change is not necessary for the next commit. I think it is ok to drop this change.

File: src/resolve/resolved-dns-synthesize.c [hunk 2]
```diff
@@ -322,8 +322,24 @@ static int synthesize_gateway_rr(Manager *m, const DnsResourceKey *key, int ifin
...
+                if (n == 0) { /* == 0 means we have no gateway */
+                        /* See if there's a gateway on the other protocol */
+                        if (af == AF_INET)
+                                n = local_gateways(m->rtnl, ifindex, AF_INET6, NULL);
+                        else {
+                                assert(af == AF_INET6);
+                                n = local_gateways(m->rtnl, ifindex, AF_INET, NULL);
+                        }
+                        if (n < 0)
+                                return n;
```
Comment (yuwata) [inline src/resolve/resolved-dns-synthesize.c]: I guess, `=` is missing:
```
if (n <= 0)
        return n;
```

**Review by yuwata — CHANGES_REQUESTED** [review]

File: src/shared/local-addresses.c (hunk 1 above)
Comment (poettering) [reply src/shared/local-addresses.c]: indeed. this makes no sense. will drop.

File: src/resolve/resolved-dns-synthesize.c (hunk 2 above)
Comment (poettering) [reply src/resolve/resolved-dns-synthesize.c]: you are absolutely right.

**Comment by poettering** [comment]
Fixed the issues @yuwata found! PTAL!

**Comment by yuwata** [comment]
LGTM.

**Pull request merged**